### PR TITLE
PostgreSQL quota manager and storage backend

### PR DESCRIPTION
**Pull request by robstradling**
This PR is based on Trillian's existing MySQL quota manager and storage backend.  The first several commits were auto-generated by [this script](https://gist.github.com/robstradling/e6685c10534ca21bb10b2871c8a154c0), which forked the existing MySQL code into different directories (whilst preserving the git history) and then did a bunch of search'n'replacing to switch from the [database/sql](https://pkg.go.dev/database/sql) interface to the [jackc/pgx](https://pkg.go.dev/github.com/jackc/pgx/v5) interface.

Improving performance is my main reason for using the pgx interface directly.  In particular, the pgx interface has allowed me to use PostgreSQL's COPY interface for fast bulk-upserts.

My motivations for putting together this PR are that (1) I and my colleagues at Sectigo have a fair amount of experience with PostgreSQL, but almost no experience with MySQL/MariaDB; and (2) we suffered a [CT log failure earlier this year](https://groups.google.com/a/chromium.org/g/ct-policy/c/038B7F4g8cU/m/KsOJaEhnBgAJ) due to MariaDB corruption after disk space exhaustion, and we are confident that PostgreSQL would not have broken under the same circumstances.

### Checklist

- [x] I have updated the [CHANGELOG](CHANGELOG.md).
- [x] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

### CODE REVIEW

File: storage/postgresql/log_storage.go
Comment (mhutchinson) [inline storage/postgresql/log_storage.go:55]: Given enough time I can probably chip away at the layers here to understand what's going on. But it strikes me that what would be super handy is a high level guide in a README.md in this directory that lays out the approach you're taking. If someone were to come and look at this code, how should they start to understand what goes in the temporary tables, the real tables, or the stored procedures? Are the temporary tables bound to a single transaction? Why are stored procedures used rather than direct queries?

I can guess answers to the above with a reasonable degree of accuracy I expect. But I'd feel much more at home if I didn't have to guess 😄  I don't expect this to be a massive tome. Just a couple of paragraphs to explain the approach and perspective you had in mind when this was written will be ace.

**Review by mhutchinson — COMMENTED** [review]
Looking great @robstradling. I've reviewed up to the meaty bits. Looking good so far. A few nitty comments, and then a request for some brief docs to help me understand the flows in the storage layers. Once that's done (and I've had my lunch) I can take a look over the remaining files.

File: .github/workflows/test_pgdb.yaml
Comment (robstradling) [reply .github/workflows/test_pgdb.yaml:68]: Makes sense in principle.  However, doing this (see https://github.com/robstradling/trillian/pull/4) causes the unit tests to fail silently...
```
	127.0.0.1:5432 (localhost): server error: FATAL: database "defaultdb" does not exist (SQLSTATE 3D000)
    postgresql_quota_test.go:118: Skipping test as PostgreSQL not available
```

The defaultdb database is created by `scripts/resetpgdb.sh`, which is executed at the beginning of the integration tests.  The unit tests benefit from this, but only if they run after the integration tests.

I copied this integration test behaviour from the existing MySQL/MariaDB and CRDB code.  See [here](https://github.com/google/trillian/pull/3644/files#diff-467c235bab405c446eead2b9ddf559737f30430bc74fdf133571ff43a871898f).

I think I'd rather leave this as-is, if that's OK.

File: cmd/trillian_log_server/main.go
Comment (robstradling) [reply cmd/trillian_log_server/main.go:52]: Here's an idea: https://github.com/robstradling/trillian/pull/5

File: quota/postgresqlqm/postgresql_quota.go
Comment (robstradling) [reply quota/postgresqlqm/postgresql_quota.go]: Addressed by https://github.com/google/trillian/pull/3644/commits/651b3b58fc362c6c10cb28a04e94bc1563e0f81c.

File: storage/postgresql/admin_storage.go
Comment (robstradling) [reply storage/postgresql/admin_storage.go]: https://github.com/google/trillian/pull/3644/commits/72964d20a0db33b1be5715638a2de4b688522923 removes the PublicKey and PrivateKey fields.

File: storage/postgresql/admin_storage.go
Comment (robstradling) [reply storage/postgresql/admin_storage.go]: https://github.com/google/trillian/pull/3644/commits/72964d20a0db33b1be5715638a2de4b688522923 removes both of these fields.

File: .github/workflows/test_pgdb.yaml
Comment (mhutchinson) [reply .github/workflows/test_pgdb.yaml:68]: Yeah, I'm just too old skool and read "unit tests" to mean "no external dependencies". Trillian has a more liberal definition of unit tests and you're following the trend. Let's leave as is.

File: quota/postgresqlqm/postgresql_quota.go
Comment (mhutchinson) [reply quota/postgresqlqm/postgresql_quota.go]: Ace.

File: storage/postgresql/admin_storage.go
Comment (mhutchinson) [reply storage/postgresql/admin_storage.go]: Good work!

File: storage/postgresql/admin_storage.go
Comment (robstradling) [reply storage/postgresql/admin_storage.go]: Addressed by https://github.com/google/trillian/pull/3644/commits/917f362c70edeb42a671b1efee99717eb82512ff.

File: quota/postgresqlqm/postgresql_quota.go
Comment (robstradling) [reply quota/postgresqlqm/postgresql_quota.go]: FWIW, this comment was copied directly from `quota/mysqlqm/mysql_quota.go`.
I'm not familiar with any of the quota managers that offer features that this one doesn't, but I've added text (based heavily on your comment) in https://github.com/google/trillian/pull/3644/commits/32096c41d0f15b9a71fe3b96e6d93aacc7b55bcd.

File: storage/postgresql/log_storage.go
Comment (robstradling) [reply storage/postgresql/log_storage.go:55]: Addressed by https://github.com/google/trillian/pull/3644/commits/8c40e33a7ad92cf11b83af87e2d2d405785036b6.

Comment (mhutchinson) [reply storage/postgresql/log_storage.go:55]: Cheers!

File: storage/postgresql/admin_storage.go
Comment (mhutchinson) [reply storage/postgresql/admin_storage.go]: Good work!

File: quota/postgresqlqm/postgresql_quota.go
Comment (mhutchinson) [reply quota/postgresqlqm/postgresql_quota.go]: Thanks for doing this.

File: storage/postgresql/queue.go
Comment (mhutchinson) [inline storage/postgresql/queue.go:2]: You're the captain of this code, so you should strongly consider which implementation of this queue you want, and then delete the other one. If you think there's a good reason to keep both, then you have my blessing. But in reality, I suspect only one of these implementations will ever be used so best to make it a good one.

File: storage/postgresql/schema/storage.sql
Comment (mhutchinson) [inline storage/postgresql/schema/storage.sql:46]: Pretty sure this is not used now.

File: storage/postgresql/schema/storage.sql
Comment (mhutchinson) [inline storage/postgresql/schema/storage.sql:62]: If my other comments about revisions were true, then this can be removed. And `TreeHeadRevisionIdx` can go.

**Review by mhutchinson — COMMENTED** [review]
Looking good. The fact that this is tested and passes the tests means that I'm happy with it. Treat my comments here as largely optional if you want the opportunity to reduce the complexity before you bring up any production logs using this. Apart from the error in CODEOWNERS. That will need to be fixed :-)

File: CODEOWNERS
Comment (robstradling) [reply CODEOWNERS]: Works for me.  Addressed by https://github.com/google/trillian/pull/3644/commits/6e3eb86c203c2b92bc59f61c70289d826522530f.

File: storage/postgresql/admin_storage.go
Comment (robstradling) [reply storage/postgresql/admin_storage.go]: All TODOs transferred to my name by https://github.com/google/trillian/pull/3644/commits/d41757ea5efca3f6282d5e6849546bc29b0bc6ca.

File: storage/postgresql/provider.go
Comment (robstradling) [reply storage/postgresql/provider.go]: All TODOs transferred to my name by https://github.com/google/trillian/pull/3644/commits/d41757ea5efca3f6282d5e6849546bc29b0bc6ca.

Comment (mhutchinson) [reply storage/postgresql/provider.go]: Ace, ta!

File: storage/postgresql/queue.go
Comment (robstradling) [reply storage/postgresql/queue.go:2]: Thanks.  I'm going with `queue_batching.go` then (although I'll rename it to `queue.go`).  Looks like this provides another 4x speed-up!!

Comment (robstradling) [reply storage/postgresql/queue.go]: Addressed by https://github.com/google/trillian/pull/3644/commits/378779f016e2e829f1abd1011c9a85cd4f23e466.

File: storage/postgresql/schema/storage.sql
Comment (robstradling) [reply storage/postgresql/schema/storage.sql]: Addressed by https://github.com/google/trillian/pull/3644/commits/8102b6ade7aa508dd2ae255dd5e721b74f522541.

File: storage/postgresql/tree_storage.go
Comment (robstradling) [reply storage/postgresql/tree_storage.go]: Addressed by https://github.com/google/trillian/pull/3644/commits/be9618289c8890e98a0aaf68c6f9bc20636343d6.

File: storage/postgresql/log_storage.go
Comment (robstradling) [reply storage/postgresql/log_storage.go]: Addressed by https://github.com/google/trillian/pull/3644/commits/be9618289c8890e98a0aaf68c6f9bc20636343d6.

File: storage/postgresql/schema/storage.sql
Comment (robstradling) [reply storage/postgresql/schema/storage.sql]: Addressed by https://github.com/google/trillian/pull/3644/commits/be9618289c8890e98a0aaf68c6f9bc20636343d6.

**Review by mhutchinson — APPROVED** [review]
Woohoo 🎉

**Comment by robstradling** [comment]
Thanks @roger2hk.  Adding postgres support for your CTFE 'extra data' storage saving feature is next on my list.  :-)

**Pull request merged**
